components/batch-system/src/batch.rs

```diff
@@ -303,30 +284,53 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
                     batch.release_control(&self.router.control_box, len);
                 }
             }
-            if !batch.normals.is_empty() {
-                for (i, p) in batch.normals.iter_mut().enumerate() {
-                    let len = self.handler.handle_normal(p);
-                    batch.counters[i] += 1;
-                    if p.is_stopped() {
-                        reschedule_fsms.push((i, ReschedulePolicy::Remove));
-                    } else {
-                        if batch.counters[i] > 3 {
-                            hot_fsm_count += 1;
-                            // We should only reschedule a half of the hot regions, otherwise,
-                            // it's possible all the hot regions are fetched in a batch the
-                            // next time.
-                            if hot_fsm_count % 2 == 0 {
-                                reschedule_fsms.push((i, ReschedulePolicy::Schedule));
-                                continue;
-                            }
-                        }
-                        if let Some(l) = len {
-                            reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+
+            let mut hot_fsm_count = 0;
+            for (i, p) in batch.normals.iter_mut().enumerate() {
+                let len = self.handler.handle_normal(p);
+                batch.counters[i] += 1;
+                if p.is_stopped() {
+                    reschedule_fsms.push((i, ReschedulePolicy::Remove));
+                } else {
+                    if batch.counters[i] > 3 {
+                        hot_fsm_count += 1;
+                        // We should only reschedule a half of the hot regions, otherwise,
+                        // it's possible all the hot regions are fetched in a batch the
+                        // next time.
+                        if hot_fsm_count % 2 == 0 {
+                            reschedule_fsms.push((i, ReschedulePolicy::Schedule));
+                            continue;
                         }
                     }
+                    if let Some(l) = len {
+                        reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+                    }
+                }
+            }
+            let mut fsm_cnt = batch.normals.len();
+            while batch.normals.len() < max_batch_size {
+                if let Ok(fsm) = self.fsm_receiver.try_recv() {
+                    run = batch.push(fsm);
+                }
+                // If we receive a ControlFsm, break this cycle and call `end`. Because ControlFsm
+                // may change state of the handler, we shall deal with it immediately after
+                // calling `begin` of `Handler`.
+                if !run || fsm_cnt >= batch.normals.len() {
```

Comment:
The `fsm_cnt >= batch.normals.len()` looks weird.
It means `try_recv` return `TryRecvError::Empty` result. Is it right?

Reply:
No, it is also possible caused by receiving a `ControlFsm`.

Reply:
I don't understand how this can show whether a `ControlFsm` is received, seems `fsm_cnt` and `batch.normals.len()` would not be changed and `run` would be `true`.

Reply:
ControlFsm will never be pushed into `normals`

Reply:
How is `fsm_cnt` increased?